# 2019 Annual Developer Survey Analysis

In [1]:
import pandas as pd

In [2]:
#convert "survey_results_public.csv" to a pandas dataframe
df = pd.read_csv("survey_results_public.csv")

### First Five Rows of the Dataframe

In [3]:
#print the first five rows of the dataframe
df.head()

Respondent                                         MainBranch Hobbyist  \
0           1             I am a student who is learning to code      Yes   
1           2             I am a student who is learning to code       No   
2           3  I am not primarily a developer, but I write co...      Yes   
3           4                     I am a developer by profession       No   
4           5                     I am a developer by profession      Yes   

                  OpenSourcer  \
0                       Never   
1     Less than once per year   
2                       Never   
3                       Never   
4  Once a month or more often   

                                          OpenSource  \
0  The quality of OSS and closed source software ...   
1  The quality of OSS and closed source software ...   
2  The quality of OSS and closed source software ...   
3  The quality of OSS and closed source software ...   
4  OSS is, on average, of HIGHER quality than pro...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   
2                      Employed full-time                Thailand   
3                      Employed full-time           United States   
4                      Employed full-time                 Ukraine   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   
2              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
3              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  ...  \
0                                                NaN  ...   
1                                                NaN  ...   
2                      Web development or web design  ...   
3  Computer science, computer engineering, or sof...  ...   
4  Computer science, computer engineering, or sof...  ...   

                             WelcomeChange  \
0  Just as welcome now as I felt last year   
1  Just as welcome now as I felt last year   
2  Just as welcome now as I felt last year   
3  Just as welcome now as I felt last year   
4  Just as welcome now as I felt last year   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  14.0    Man    No   
1  Tech articles written by other developers;Indu...  19.0    Man    No   
2  Tech meetups or events in your area;Courses on...  28.0    Man    No   
3  Tech articles written by other developers;Indu...  22.0    Man    No   
4  Tech meetups or events in your area;Courses on...  30.0    Man    No   

                 Sexuality                                 Ethnicity  \
0  Straight / Heterosexual                                       NaN   
1  Straight / Heterosexual                                       NaN   
2  Straight / Heterosexual                                       NaN   
3  Straight / Heterosexual              White or of European descent   
4  Straight / Heterosexual  White or of European descent;Multiracial   

  Dependents           SurveyLength                  SurveyEase  
0         No  Appropriate in length  Neither easy nor difficult  
1         No  Appropriate in length  Neither easy nor difficult  
2        Yes  Appropriate in length  Neither easy nor difficult  
3         No  Appropriate in length                        Easy  
4         No  Appropriate in length                        Easy  

[5 rows x 85 columns]

In [4]:
pd.set_option("display.max_row", None)

In [5]:
social_media_counts = df["SocialMedia"].value_counts()
print(social_media_counts)

Reddit                      14374
YouTube                     13830
WhatsApp                    13347
Facebook                    13178
Twitter                     11398
Instagram                    6261
I don't use social media     5554
LinkedIn                     4501
WeChat 微信                     667
Snapchat                      628
VK ВКонта́кте                 603
Weibo 新浪微博                     56
Youku Tudou 优酷                 21
Hello                          19
Name: SocialMedia, dtype: int64


### Number of people using each type of social media:
- Reddit: 14374
- YouTube: 13830
- WhatsApp: 13347
- Facebook: 13178
- Twitter: 11398
- Instagram: 6261
- I don't use social media: 5554
- LinkedIn: 4501
- WeChat 微信: 667
- Snapchat: 628
- VK ВКонта́кте: 603
- Weibo 新浪微博: 56
- Youku Tudou 优酷: 21
- Hello: 19

In [6]:
#Dimensions of df
df.shape

(88883, 85)

#### There are 88885 rows, and 85 columns.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88883 entries, 0 to 88882
Data columns (total 85 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Respondent              88883 non-null  int64  
 1   MainBranch              88331 non-null  object 
 2   Hobbyist                88883 non-null  object 
 3   OpenSourcer             88883 non-null  object 
 4   OpenSource              86842 non-null  object 
 5   Employment              87181 non-null  object 
 6   Country                 88751 non-null  object 
 7   Student                 87014 non-null  object 
 8   EdLevel                 86390 non-null  object 
 9   UndergradMajor          75614 non-null  object 
 10  EduOther                84260 non-null  object 
 11  OrgSize                 71791 non-null  object 
 12  DevType                 81335 non-null  object 
 13  YearsCode               87938 non-null  object 
 14  Age1stCode              87634 non-null

In [8]:
pd.set_option("display.max_row", 85)
pd.set_option("display.max_columns", 85)

In [9]:
df.head()

Respondent                                         MainBranch Hobbyist  \
0           1             I am a student who is learning to code      Yes   
1           2             I am a student who is learning to code       No   
2           3  I am not primarily a developer, but I write co...      Yes   
3           4                     I am a developer by profession       No   
4           5                     I am a developer by profession      Yes   

                  OpenSourcer  \
0                       Never   
1     Less than once per year   
2                       Never   
3                       Never   
4  Once a month or more often   

                                          OpenSource  \
0  The quality of OSS and closed source software ...   
1  The quality of OSS and closed source software ...   
2  The quality of OSS and closed source software ...   
3  The quality of OSS and closed source software ...   
4  OSS is, on average, of HIGHER quality than pro...   

                               Employment                 Country  \
0  Not employed, and not looking for work          United Kingdom   
1      Not employed, but looking for work  Bosnia and Herzegovina   
2                      Employed full-time                Thailand   
3                      Employed full-time           United States   
4                      Employed full-time                 Ukraine   

          Student                                            EdLevel  \
0              No                          Primary/elementary school   
1  Yes, full-time  Secondary school (e.g. American high school, G...   
2              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
3              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4              No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  \
0                                                NaN   
1                                                NaN   
2                      Web development or web design   
3  Computer science, computer engineering, or sof...   
4  Computer science, computer engineering, or sof...   

                                            EduOther  \
0  Taught yourself a new language, framework, or ...   
1  Taken an online course in programming or softw...   
2  Taught yourself a new language, framework, or ...   
3  Taken an online course in programming or softw...   
4  Taken an online course in programming or softw...   

                    OrgSize  \
0                       NaN   
1                       NaN   
2      100 to 499 employees   
3      100 to 499 employees   
4  10,000 or more employees   

                                             DevType YearsCode Age1stCode  \
0                                                NaN         4         10   
1  Developer, desktop or enterprise applications;...       NaN         17   
2  Designer;Developer, back-end;Developer, front-...         3         22   
3                              Developer, full-stack         3         16   
4  Academic researcher;Developer, desktop or ente...        16         14   

       YearsCodePro           CareerSat                 JobSat  \
0               NaN                 NaN                    NaN   
1               NaN                 NaN                    NaN   
2                 1  Slightly satisfied     Slightly satisfied   
3  Less than 1 year      Very satisfied     Slightly satisfied   
4                 9   Very dissatisfied  Slightly dissatisfied   

               MgrIdiot  MgrMoney   MgrWant  \
0                   NaN       NaN       NaN   
1                   NaN       NaN       NaN   
2  Not at all confident  Not sure  Not sure   
3        Very confident        No  Not sure   
4    Somewhat confident       Yes        No   

                                             JobSeek          LastHireDate  \
0                                                NaN                   NaN

In [10]:
schema_df = pd.read_csv("survey_results_schema.csv")

In [11]:
schema_df.head()

Column                                       QuestionText
0   Respondent  Randomized respondent ID number (not in order ...
1   MainBranch  Which of the following options best describes ...
2     Hobbyist                            Do you code as a hobby?
3  OpenSourcer        How often do you contribute to open source?
4   OpenSource  How do you feel about the quality of open sour...

In [12]:
schema_df.set_index("Column", inplace=True)

In [13]:
schema_df

QuestionText
Column                                                                   
Respondent              Randomized respondent ID number (not in order ...
MainBranch              Which of the following options best describes ...
Hobbyist                                          Do you code as a hobby?
OpenSourcer                   How often do you contribute to open source?
OpenSource              How do you feel about the quality of open sour...
Employment              Which of the following best describes your cur...
Country                         In which country do you currently reside?
Student                 Are you currently enrolled in a formal, degree...
EdLevel                 Which of the following best describes the high...
UndergradMajor          What was your main or most important field of ...
EduOther                Which of the following types of non-degree edu...
OrgSize                 Approximately how many people are employed by ...
DevType                 Which of the following describe you? Please se...
YearsCode               Including any education, how many years have y...
Age1stCode              At what age did you write your first line of c...
YearsCodePro            How many years have you coded professionally (...
CareerSat               Overall, how satisfied are you with your caree...
JobSat                  How satisfied are you with your current job? (...
MgrIdiot                How confident are you that your manager knows ...
MgrMoney                Do you believe that you need to be a manager t...
MgrWant                 Do you want to become a manager yourself in th...
JobSeek                 Which of the following best describes your cur...
LastHireDate            When was the last time that you took a job wit...
LastInt                 In your most recent successful job interview (...
FizzBuzz                Have you ever been asked to solve FizzBuzz in ...
JobFactors              Imagine that you are deciding between two job ...
ResumeUpdate            Think back to the last time you updated your r...
CurrencySymbol          Which currency do you use day-to-day? If your ...
CurrencyDesc            Which currency do you use day-to-day? If your ...
CompTotal               What is your current total compensation (salar...
CompFreq                 Is that compensation weekly, monthly, or yearly?
ConvertedComp           Salary converted to annual USD salaries using ...
WorkWeekHrs              On average, how many hours per week do you work?
WorkPlan                          How structured or planned is your work?
WorkChallenge           Of these options, what are your greatest chall...
WorkRemote                                How often do you work remotely?
WorkLoc                                   Where would you prefer to work?
ImpSyn                  For the specific work you do, and the years of...
CodeRev                          Do you review code as part of your work?
CodeRevHrs              On average, how many hours per week do you spe...
UnitTests               Does your company regularly employ unit tests ...
PurchaseHow             How does your company make decisions about pur...
PurchaseWhat            What level of influence do you, personally, ha...
LanguageWorkedWith      Which of the following programming, scripting,...
LanguageDesireNextYear  Which of the following programming, scripting,...
DatabaseWorkedWith      Which of the following database environments h...
DatabaseDesireNextYear  Which of the following database environments h...
PlatformWorkedWith      Which of the following platforms have you done...
PlatformDesireNextYear  Which of the following platforms have you done...
WebFrameWorkedWith      Which of the following web frameworks have you...
WebFrameDesireNextYear  Which of the following web frameworks have you...
MiscTechWorkedWith      Which of the following other frameworks, libra...
MiscTechDesireNextYear  Which of the following other frameworks, 

In [14]:
schema_df = schema_df.sort_index()

In [15]:
schema_df

QuestionText
Column                                                                   
Age                     What is your age (in years)? If you prefer not...
Age1stCode              At what age did you write your first line of c...
BetterLife              Do you think people born today will have a bet...
BlockchainIs            Blockchain / cryptocurrency technology is prim...
BlockchainOrg           How is your organization thinking about or imp...
CareerSat               Overall, how satisfied are you with your caree...
CodeRev                          Do you review code as part of your work?
CodeRevHrs              On average, how many hours per week do you spe...
CompFreq                 Is that compensation weekly, monthly, or yearly?
CompTotal               What is your current total compensation (salar...
Containers              How do you use containers (Docker, Open Contai...
ConvertedComp           Salary converted to annual USD salaries using ...
Country                         In which country do you currently reside?
CurrencyDesc            Which currency do you use day-to-day? If your ...
CurrencySymbol          Which currency do you use day-to-day? If your ...
DatabaseDesireNextYear  Which of the following database environments h...
DatabaseWorkedWith      Which of the following database environments h...
Dependents              Do you have any dependents (e.g., children, el...
DevEnviron              Which development environment(s) do you use re...
DevType                 Which of the following describe you? Please se...
EdLevel                 Which of the following best describes the high...
EduOther                Which of the following types of non-degree edu...
Employment              Which of the following best describes your cur...
EntTeams                Have you ever used Stack Overflow for Enterpri...
Ethnicity               Which of the following do you identify as? Ple...
Extraversion              Do you prefer online chat or IRL conversations?
FizzBuzz                Have you ever been asked to solve FizzBuzz in ...
Gender                  Which of the following do you currently identi...
Hobbyist                                          Do you code as a hobby?
ITperson                 Are you the "IT support person" for your family?
ImpSyn                  For the specific work you do, and the years of...
JobFactors              Imagine that you are deciding between two job ...
JobSat                  How satisfied are you with your current job? (...
JobSeek                 Which of the following best describes your cur...
LanguageDesireNextYear  Which of the following programming, scripting,...
LanguageWorkedWith      Which of the following programming, scripting,...
LastHireDate            When was the last time that you took a job wit...
LastInt                 In your most recent successful job interview (...
MainBranch              Which of the following options best describes ...
MgrIdiot                How confident are you that your manager knows ...
MgrMoney                Do you believe that you need to be a manager t...
MgrWant                 Do you want to become a manager yourself in th...
MiscTechDesireNextYear  Which of the following other frameworks, libra...
MiscTechWorkedWith      Which of the following other frameworks, libra...
OffOn                         Have you tried turning it off and on again?
OpSys                   What is the primary operating system in which ...
OpenSource              How do you feel about the quality of open sour...
OpenSourcer                   How often do you contribute to open source?
OrgSize                 Approximately how many people are employed by ...
PlatformDesireNextYear  Which of the following platforms have you done...
PlatformWorkedWith      Which of the following platforms have you done...
PurchaseHow             How does your company make decisions about pur...
PurchaseWhat            What level of influence do you, personall

In [16]:
schema_df.loc["SOJobs"]

QuestionText    Have you ever used or visited Stack Overflow J...
Name: SOJobs, dtype: object

In [17]:
schema_df.loc["MgrIdiot"]

QuestionText    How confident are you that your manager knows ...
Name: MgrIdiot, dtype: object

In [18]:
df.describe().T

count          mean           std  min      25%      50%  \
Respondent     88883.0  4.444200e+04  2.565846e+04  1.0  22221.5  44442.0   
CompTotal      55945.0  5.519014e+11  7.331926e+13  0.0  20000.0  62000.0   
ConvertedComp  55823.0  1.271107e+05  2.841523e+05  0.0  25777.5  57287.0   
WorkWeekHrs    64503.0  4.212720e+01  3.728761e+01  1.0     40.0     40.0   
CodeRevHrs     49790.0  5.084308e+00  5.513931e+00  0.0      2.0      4.0   
Age            79210.0  3.033670e+01  9.178390e+00  1.0     24.0     29.0   

                     75%           max  
Respondent      66662.50  8.888300e+04  
CompTotal      120000.00  1.000000e+16  
ConvertedComp  100000.00  2.000000e+06  
WorkWeekHrs        44.75  4.850000e+03  
CodeRevHrs          6.00  9.900000e+01  
Age                35.00  9.900000e+01

In [19]:
hobbyist_counts = df["Hobbyist"].value_counts()
print(hobbyist_counts)

Yes    71257
No     17626
Name: Hobbyist, dtype: int64


### 71,257 People code as a hobby

In [20]:
df.loc[0]

Respondent                                                                1
MainBranch                           I am a student who is learning to code
Hobbyist                                                                Yes
OpenSourcer                                                           Never
OpenSource                The quality of OSS and closed source software ...
Employment                           Not employed, and not looking for work
Country                                                      United Kingdom
Student                                                                  No
EdLevel                                           Primary/elementary school
UndergradMajor                                                          NaN
EduOther                  Taught yourself a new language, framework, or ...
OrgSize                                                                 NaN
DevType                                                                 NaN
YearsCode   

In [21]:
df["Hobbyist"].loc[[2171, 8800]]

2171     No
8800    Yes
Name: Hobbyist, dtype: object

In [22]:
df["Hobbyist"].loc[8771:8802]

8771    Yes
8772     No
8773     No
8774    Yes
8775    Yes
8776    Yes
8777     No
8778    Yes
8779    Yes
8780    Yes
8781    Yes
8782    Yes
8783     No
8784    Yes
8785     No
8786    Yes
8787    Yes
8788    Yes
8789     No
8790     No
8791    Yes
8792    Yes
8793    Yes
8794    Yes
8795    Yes
8796    Yes
8797    Yes
8798    Yes
8799     No
8800    Yes
8801    Yes
8802    Yes
Name: Hobbyist, dtype: object

In [23]:
df[["Hobbyist", "OpenSourcer", "OpenSource", "Employment", "Country", "Student", "EdLevel"]].loc[8880:8882]

Hobbyist                 OpenSourcer  \
8880      Yes                       Never   
8881      Yes                       Never   
8882      Yes  Once a month or more often   

                                             OpenSource  \
8880  The quality of OSS and closed source software ...   
8881  The quality of OSS and closed source software ...   
8882  The quality of OSS and closed source software ...   

                                             Employment         Country  \
8880                                 Employed full-time  United Kingdom   
8881                                 Employed full-time          Brazil   
8882  Independent contractor, freelancer, or self-em...           India   

             Student                                            EdLevel  
8880              No           Bachelor’s degree (BA, BS, B.Eng., etc.)  
8881  Yes, full-time  Some college/university study without earning ...  
8882              No           Bachelor’s degree (BA, BS, B.Eng., etc.)

In [24]:
df.set_index("Respondent", inplace=True)

In [25]:
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

In [26]:
high_salary = (df["ConvertedComp"] > 7000)

In [27]:
df.loc[high_salary, ['Country', 'LanguageWorkedWith','ConvertedComp']].sort_values('ConvertedComp', ascending=False).head(100)

Country                                 LanguageWorkedWith  \
Respondent                                                                     
39044       United States                     C;C++;JavaScript;Python;R;Rust   
44522       United States                                         Ruby;Swift   
42005       United States   HTML/CSS;JavaScript;Objective-C;Python;SQL;Swift   
13802       United States  HTML/CSS;Java;JavaScript;SQL;TypeScript;Other(s):   
63965       United States  Bash/Shell/PowerShell;C#;Go;HTML/CSS;JavaScrip...   
...                   ...                                                ...   
42292       United States            HTML/CSS;Java;JavaScript;SQL;TypeScript   
72189       United States  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
44552       United States       Bash/Shell/PowerShell;Java;Python;Ruby;Scala   
15692       United States                                  C#;JavaScript;SQL   
48022       United States                                    Java;JavaScript   

            ConvertedComp  
Respondent                 
39044           2000000.0  
44522           2000000.0  
42005           2000000.0  
13802           2000000.0  
63965           2000000.0  
...                   ...  
42292           2000000.0  
72189           2000000.0  
44552           2000000.0  
15692           2000000.0  
48022           2000000.0  

[100 rows x 3 columns]

In [28]:
def python_string(title):
    if "python" in str(title).lower():
        return True
    else:
        return False

In [29]:
df[df["LanguageWorkedWith"].map(lambda x: python_string(x))]["LanguageWorkedWith"]

Respondent
1                          HTML/CSS;Java;JavaScript;Python
2                                      C++;HTML/CSS;Python
4                                      C;C++;C#;Python;SQL
5              C++;HTML/CSS;Java;JavaScript;Python;SQL;VBA
8        Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
                               ...                        
84539    Bash/Shell/PowerShell;C;C++;HTML/CSS;Java;Java...
85738      Bash/Shell/PowerShell;C++;Python;Ruby;Other(s):
86566      Bash/Shell/PowerShell;HTML/CSS;Python;Other(s):
87739             C;C++;HTML/CSS;JavaScript;PHP;Python;SQL
88212                           HTML/CSS;JavaScript;Python
Name: LanguageWorkedWith, Length: 36443, dtype: object

In [30]:
df = df.rename(columns={"ConvertedComp": "Salary_in_USD"})

In [31]:
df.head(3)

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   

                        OpenSourcer  \
Respondent                            
1                             Never   
2           Less than once per year   
3                             Never   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                           UndergradMajor  \
Respondent                                  
1                                     NaN   
2                                     NaN   
3           Web development or web design   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   

                         OrgSize  \
Respondent                         
1                            NaN   
2                            NaN   
3           100 to 499 employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   

           Age1stCode YearsCodePro           CareerSat              JobSat  \
Respondent                                                                   
1                  10          NaN                 NaN                 NaN   
2                  17          NaN                 NaN                 NaN   
3                  22            1  Slightly satisfied  Slightly satisfied   

                        MgrIdiot  MgrMoney   MgrWant  \
Respondent                                             
1                            NaN       NaN       NaN   
2                            NaN       NaN       NaN   
3           Not at all confident  Not sure  Not sure   

                                                      JobSeek  \
Respondent                                                      
1                                                         NaN   
2                             I am actively looking for a job   
3           I’m not actively looking, but I am open to new...   

                    LastHireDate                              LastInt  \
Respondent                                                              
1                            NaN                                  NaN   
2           I've never had a job   

In [32]:
df["Hobbyist"] = df["Hobbyist"].map({"Yes": "True", "No": "False"})

In [33]:
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code     True   
2                      I am a student who is learning to code    False   
3           I am not primarily a developer, but I write co...     True   
4                              I am a developer by profession    False   
5                              I am a developer by profession     True   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

In [34]:
hobbyist_counts = df["Hobbyist"].value_counts()
print(hobbyist_counts)

True     71257
False    17626
Name: Hobbyist, dtype: int64


In [35]:
df = df.sort_values("Country")

In [50]:
df.head()

MainBranch Hobbyist              OpenSourcer  \
Respondent                                                                     
39258       I am a developer by profession     True  Less than once per year   
63129       I am a developer by profession     True                    Never   
85715       I am a developer by profession    False  Less than once per year   
50767       I am a developer by profession    False  Less than once per year   
2782                                   NaN     True  Less than once per year   

                                                   OpenSource  \
Respondent                                                      
39258       OSS is, on average, of LOWER quality than prop...   
63129       OSS is, on average, of HIGHER quality than pro...   
85715       The quality of OSS and closed source software ...   
50767                                                     NaN   
2782        OSS is, on average, of HIGHER quality than pro...   

                                                   Employment      Country  \
Respondent                                                                   
39258                                      Employed full-time  Afghanistan   
63129                                      Employed full-time  Afghanistan   
85715                                      Employed full-time  Afghanistan   
50767       Independent contractor, freelancer, or self-em...  Afghanistan   
2782                                                      NaN  Afghanistan   

                   Student                                   EdLevel  \
Respondent                                                             
39258                   No  Bachelor’s degree (BA, BS, B.Eng., etc.)   
63129       Yes, full-time    I never completed any formal education   
85715                   No  Bachelor’s degree (BA, BS, B.Eng., etc.)   
50767       Yes, part-time                          Associate degree   
2782                    No                                       NaN   

                                               UndergradMajor  \
Respondent                                                      
39258       Computer science, computer engineering, or sof...   
63129                                                     NaN   
85715       Information systems, information technology, o...   
50767       A health science (ex. nursing, pharmacy, radio...   
2782                                 I never declared a major   

                                                     EduOther  \
Respondent                                                      
39258       Taken an online course in programming or softw...   
63129       Taken an online course in programming or softw...   
85715       Received on-the-job training in software devel...   
50767       Completed an industry certification program (e...   
2782        Taken an online course in programming or softw...   

                                                      OrgSize  \
Respondent                                                      
39258                                    100 to 499 employees   
63129       Just me - I am a freelancer, sole proprietor, ...   
85715                                                     NaN   
50767                                1,000 to 4,999 employees   
2782                                                      NaN   

                                                      DevType  \
Respondent                                                      
39258       Database administrator;Designer;Developer, bac...   
63129                                                     NaN   
85715                                     Developer, back-end   
50767       Developer, desktop or enterprise applications;...   
2782                                                      NaN   

                   YearsCode            Age1stCode      YearsCodePro  \
Respondent                                                         

In [37]:
df['Country'].value_counts()[["United States", "India", "Germany", "United Kingdom", "Canada"]]

United States     20949
India              9061
Germany            5866
United Kingdom     5737
Canada             3395
Name: Country, dtype: int64

In [51]:
group_of_countries = df.groupby("Country").mean()

/var/folders/_x/23fx564x12b5hw88dh344lr40000gn/T/ipykernel_35874/3286342200.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  group_of_countries = df.groupby("Country").mean()


In [52]:
group_of_countries["Salary_in_USD"]

Country
Afghanistan                             101953.333333
Albania                                  21833.700000
Algeria                                  34924.047619
Andorra                                 160931.000000
Angola                                    7764.000000
                                            ...      
Venezuela, Bolivarian Republic of...     14581.627907
Viet Nam                                 17233.436782
Yemen                                    16909.166667
Zambia                                   10075.375000
Zimbabwe                                 34046.666667
Name: Salary_in_USD, Length: 179, dtype: float64

In [40]:
list_of_countries = ['Nepal', 'Mali', 'United States','United Kingdom','Canada','France']

### Top 5 Social Media Platforms in Nepal, Mali, United States, United Kingdom, Canada, and France

In [81]:
for x in list_of_countries:
    print(x + "'s top 10 SocialMedia","\n",
          df[df['Country']== x]['SocialMedia'].value_counts().nlargest(10),
          "\n", "************************************")

Nepal's top 10 SocialMedia 
 YouTube                     95
Facebook                    71
LinkedIn                    22
Twitter                     16
Reddit                       9
Instagram                    8
WhatsApp                     5
I don't use social media     4
WeChat 微信                    1
Name: SocialMedia, dtype: int64 
 ************************************
Mali's top 10 SocialMedia 
 WhatsApp    1
YouTube     1
Facebook    1
Name: SocialMedia, dtype: int64 
 ************************************
United States's top 10 SocialMedia 
 Reddit                      5700
Twitter                     3468
Facebook                    2844
YouTube                     2463
I don't use social media    1851
Instagram                   1652
LinkedIn                    1020
WhatsApp                     609
Snapchat                     326
WeChat 微信                     93
Name: SocialMedia, dtype: int64 
 ************************************
United Kingdom's top 10 SocialMedia 
 Redd

### Top Salary(USD) in Nepal, Mali, United States, United Kingdom, Canada, and France¶

In [80]:
for x in list_of_countries:
    print(x + "'s top 10 Salry_in_USD","\n",
          df[df['Country']== x]['Salary_in_USD'].value_counts().nlargest(10),
          "\n", "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ ")

Nepal's top 10 Salry_in_USD 
 4224.0     8
3168.0     8
2640.0     4
6324.0     4
5268.0     4
8436.0     4
3696.0     3
10548.0    3
2112.0     3
4740.0     3
Name: Salary_in_USD, dtype: int64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Mali's top 10 Salry_in_USD 
 10476.0    1
Name: Salary_in_USD, dtype: int64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
United States's top 10 Salry_in_USD 
 2000000.0    698
120000.0     481
100000.0     455
150000.0     423
80000.0      388
110000.0     369
90000.0      367
130000.0     365
140000.0     329
70000.0      317
Name: Salary_in_USD, dtype: int64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
United Kingdom's top 10 Salry_in_USD 
 65424.0      161
52339.0      146
78508.0      144
58881.0      137
45797.0      114
1000000.0    111
39254.0      106
71966.0      103
91593.0       78
85051.0       74
Name: Salary_in_USD, dtype: int64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
Canada's top 10 Salry_in_USD 
 76339.0      116
1000000.0    109
61071.0      106
68705.0       75
45803.0       

### Mean, Max, and Median Salary(USD) in Nepal, Mali, United States, United Kingdom, Canada, and France¶

In [86]:
for x in list_of_countries:
    print(x,"\n",
          df[df['Country']== x]['Salary_in_USD'].agg(["mean", "max", "median"]),
          "\n", "_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ ", "\n")

Nepal 
 mean       10434.987952
max       300000.000000
median      4224.000000
Name: Salary_in_USD, dtype: float64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

Mali 
 mean      10476.0
max       10476.0
median    10476.0
Name: Salary_in_USD, dtype: float64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

United States 
 mean      2.495463e+05
max       2.000000e+06
median    1.100000e+05
Name: Salary_in_USD, dtype: float64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

United Kingdom 
 mean      1.661825e+05
max       2.000000e+06
median    6.804100e+04
Name: Salary_in_USD, dtype: float64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

Canada 
 mean      1.340186e+05
max       2.000000e+06
median    6.870500e+04
Name: Salary_in_USD, dtype: float64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

France 
 mean        81214.779722
max       1000000.000000
median      46752.000000
Name: Salary_in_USD, dtype: float64 
 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

